packages/astro/src/core/app/index.ts

```diff
@@ -307,6 +262,62 @@ export class App {
 		}
 	}
 
+	/**
+	 * If is a known error code, try sending the according page (e.g. 404.astro / 500.astro).
+	 * This also handles pre-rendered /404 or /500 routes
+	 */
+	async #renderError(request: Request, { routeData, status, response: originalResponse }: RenderErrorOptions) {
+		const errorRouteData = matchRoute('/' + status, this.#manifestData);
+		const url = new URL(request.url);
+		if (errorRouteData) {
+			if (errorRouteData.prerender) {
+				const statusURL = new URL(`${this.#baseWithoutTrailingSlash}/${status}`, url);
+				const response = await fetch(statusURL.toString());
+				return this.#mergeResponses(response, originalResponse);
+			}
+			const finalRouteData = routeData ?? errorRouteData;
+			const mod = await this.#getModuleForRoute(errorRouteData);
+			try {
+				const newRenderContext = await this.#createRenderContext(
+					url,
+					request,
+					finalRouteData,
+					mod,
+					status
+				);
+				const page = (await mod.page()) as any;
+				const response = await tryRenderRoute(
+					'page', // this is hardcoded to ensure proper behavior for missing endpoints
+					newRenderContext,
+					this.#env,
+					page
+				) as Response;
+				return this.#mergeResponses(response, originalResponse);
+			} catch {}
+		}
+
+		const response = this.#mergeResponses(new Response(), originalResponse);
+		Reflect.set(response, responseSentSymbol, true);
+		return response;
+	}
+
+	#mergeResponses(newResponse: Response, oldResponse?: Response) {
+		if (!oldResponse) return newResponse;
+		const { status, statusText, headers } = oldResponse;
+		return new Response(newResponse.body, {
+			status,
+			statusText,
+			headers: new Headers(Array.from(headers))
+		})
+	}
```

Comment:
Not sure if this is totally needed? We had a test that confirmed that `status` and `statusText` from the original `Response` were preserved.

Trying to think through edge cases here—a bit unclear if headers from the `newResponse` should be preserved/have precedence?

Reply:
I'm not sure either but preserving them feels like the right thing to do...